packages/webapp/src/containers/Animals/Inventory/KPI/useAnimalsFilterReduxState.ts

```diff
@@ -0,0 +1,50 @@
+/*
+ *  Copyright 2024 LiteFarm.org
+ *  This file is part of LiteFarm.
+ *
+ *  LiteFarm is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  LiteFarm is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *  GNU General Public License for more details, see <https://www.gnu.org/licenses/>.
+ */
+
+import { useDispatch, useSelector } from 'react-redux';
+import produce from 'immer';
+import { AnimalsFilterKeys } from '../../../Filter/Animals/types';
+import { animalsFilterSelector, setAnimalsFilter } from '../../../filterSlice';
+import { getActiveTypeIds } from '../../../Filter/Animals/utils';
+
+export const useAnimalsFilterReduxState = () => {
+  const dispatch = useDispatch();
+  const animalsFilter = useSelector(animalsFilterSelector);
+
+  const selectedTypeIds = getActiveTypeIds(animalsFilter);
+
+  const updateSelectedTypeIds = (typeIds: string[]) => {
+    let newFilterState = produce(
+      animalsFilter[AnimalsFilterKeys.TYPE],
+      (filterState: { [key: string]: { label?: string; active: boolean } }) => {
+        for (const key in filterState) {
+          filterState[key].active = false;
+        }
+
+        for (const key of typeIds) {
+          if (!filterState[key]) {
+            filterState[key] = { active: true };
+          } else {
+            filterState[key].active = true;
+          }
+        }
+      },
+    );
+
+    dispatch(setAnimalsFilter({ ...animalsFilter, [AnimalsFilterKeys.TYPE]: newFilterState }));
+  };
```

Comment:
This works great, but could be simplified by having this function receive the type that was clicked and its state rather than the whole state object from the component that's calling it. If the Inventory component, the onClick function could do

```js
const onTypeClick = useCallback(
    (typeId: string) => {
      const isSelected = selectedTypeIds.includes(typeId);
      updateSelectedTypeIds(typeId, isSelected);
    },
    [updateSelectedTypeIds],
  );
```

And then on the hook

```js
const updateSelectedTypeIds = (typeId: string, isSelected: boolean) => {
    const newFilterState = {
      ...animalsFilter[AnimalsFilterKeys.TYPE],
      [typeId]: { active: !isSelected },
    };
    dispatch(setAnimalsFilter({ ...animalsFilter, [AnimalsFilterKeys.TYPE]: newFilterState }));
  };
```

Basically you only keep track of the entire state object on the Redux store, and let the component only pass what's changed.

Reply:
Come to think of it we probably don't need to pass `isSelected` from the component either, since the hook already has access to `selectedTypeIds`, so it could just do

```js
const updateSelectedTypeIds = (typeId: string) => {
    const isSelected = selectedTypeIds.includes(typeId);
    const newFilterState = {
      ...animalsFilter[AnimalsFilterKeys.TYPE],
      [typeId]: { active: !isSelected },
    };
    dispatch(setAnimalsFilter({ ...animalsFilter, [AnimalsFilterKeys.TYPE]: newFilterState }));
};
```

Reply:
Gosh that's SO much cleaner 😍 I don't feel like I could have come up with that solution though, no matter how long I thought about it 😅 The `active: !isSelected` is such a good way to do the toggle!!

Reply:
It's easier when you already have someone else's code to read through and think through without being too involved 😅 glad it helps!